File: src/main/ruby/com_github_leetcode/random/node.rb
class Node
  attr_accessor :val, :next, :random

  def initialize(val = 0, next_node = nil, random = nil)
    @val = val
    @next = next_node
    @random = random
  end

  def to_s
    result = []
    curr = self

    while curr
      result << "[#{curr.val}, #{curr.random ? find_index(curr.random) : 'nil'}]"
      curr = curr.next
    end

    '[' + result.join(',') + ']'
  end

  private

  def find_index(target)
    index = 0
    curr = self

    while curr&.next && curr != target
      index += 1
      curr = curr.next
    end

    index
  end
end

File: src/main/js/com_github_leetcode/node.js
class Node {
    constructor(val, next, random) {
        this.val = val ?? 0
        this.next = next ?? null
        this.random = random ?? null
    }

    toString() {
        const result = []
        let curr = this
        while (curr !== null) {
            const result2 = []
            result2.push(String(curr.val))
            if (curr.random === null) {
                result2.push('null')
            } else {
                let randomIndex = 0
                let curr2 = this
                while (curr2?.next !== null && curr2 !== curr.random) {
                    randomIndex++
                    curr2 = curr2.next
                }
                result2.push(String(randomIndex))
            }
            result.push(`[${result2.join(',')}]`)
            curr = curr.next
        }
        return `[${result.join(',')}]`
    }
};

export { Node }
